### ISSUE TITLE
[New Feature] Capturing Deadlocks

### ISSUE THREAD

**Issue by ColinDouglas92** [marcingminski/sqlwatch#31]
**Is your feature request related to a problem? Please describe.**
Capturing deadlock information will help a user identify the queries/processes, lock types, graphs, etc. involved which will assist in implementing a solution to handling the deadlock(s). 

**Describe the solution you'd like**
From a technical details perspective, a number of ways to capture the deadlock information. My suggestion would either be using an extended event or just capture what's provided in the ring buffer. As far as displaying the data, I assume we would want a visualization similar to the whoisactive tab.

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by marcingminski** [comment]
Agree. I actually have that coded already on the back of the deadlock monitor with the help of the service broker to log into a table to show on a Gantt chart. The other alternative is to use extended events but it’s often tricky to get data out of the XE file

**Comment by ColinDouglas92** [comment]
Nice! That will be a great feature! I'll continue this thread for more ideas until we land on one that hasn't been implemented or in development 😄 I'll create a new issue when that time comes. 

On the topic of blocking/locking/deadlocking, is there anything currently being implemented to capture granular lock details? To elaborate a little bit more, this would collect all of the request modes (Shared, Intent Shared, Exclusive, Intent Exclusive, etc.) for each request type (Database, Object, Page, Key) for each session lock. 

Let me know if you need additional detail or have any questions, comments, or concerns! Thanks!

**Comment by marcingminski** [comment]
It is a wider discussion that I do not have a precise answer to (yet) as we are dealing with two different problems. 

Granular lock details for running transactions are what sp_whoisactive is really good at but we do not capture it right now. We could add more columns to the sp_whoisactive table and capture more info every 1 minute. I actually got in touch with Adam Machanic regarding our use of sp_whoisactive but he is yet to come back to me. 

Blocking and Deadlocking could be done on the back of the SQL deadlock monitor and only triggered if there is blocking detected - so could be never on some systems but which would not mean they are problem free. Blocking can also be done via sp_whoisactive but we are running it every 1 minute so you could have blocking that lasts 58 seconds every minute but never capture it. I prefer deadlock monitor for this.

In the past, I have done it using custom traces and deadlock monitor. There are also extended events which we could utilise.

Another problem with this I had in the past was storage. on busy systems the transaction table can grow to tens of GBs

**Comment by ColinDouglas92** [comment]
Good points. Thanks for the response. I noticed the sp_whoisactive call with @get_locks = 1 can also take awhile to run on some systems (increasing sp_whoisactive runtime to 2+ minutes), compared to querying the dmv's directly which take just seconds on the same system. I assume this has something to do with the XML, but haven't dug into it too much. So just something to consider as well. Thanks!

**Comment by alexander-suprun** [comment]
> Agree. I actually have that coded already on the back of the deadlock monitor with the help of the service broker to log into a table to show on a Gantt chart. The other alternative is to use extended events but it’s often tricky to get data out of the XE file

I'm not sure why it has to be so complicated. Deadlock XML is already collected as a file and in memory in system_health session starting from 2012. It's even available in 2008 in memory only and the query would be a little bit different, but it's totally possible to extract that kind of information without setting up anything else on the server.
At the same time blocker information was never a part of system_health, so I'm not sure why usp_sqlwatch_logger_xes_blockers procedure is trying to extract it from there. It will never work unless custom SQLWATCH_blockers session is enabled. Also "blocked process threshold (s)" must be configured for this XEvent to start collecting any data.

**Comment by marcingminski** [comment]
Absolutely does not have to be. This was a legacy code that I had already but never made into public release. Collecting deadlocks will follow the same approach as existing extended events collectors. The only problem with the default system_health is that it can get big and become expensive and slow to parse, which is one of the reasons for custom SQLWATCH sessions. I am also considering using [XesSmartTarget](https://github.com/spaghettidba/XESmartTarget). Good point about blockers not being logged in system_health. This and the requirement for blocked process threshold will be in docs when 2.0 is released. Thank you for your input.

**Comment by alexander-suprun** [comment]
I've never experienced an issue that it's too slow to parse system_health. First of all it's very limited, 4 files 5MB each for event file and 4mb or no more than 5000 event for ring buffer. The only problem is that if you don't read it often enough you may loose some events.

It's an interesting concept but I'm not sure how you can utilize XesSmartTarget for this project. You basically have to install and continuously run some additional exe file on the server. Unless I don't understand something.

**Comment by TheRockStarDBA** [comment]
How about using Event Notifications - https://gist.github.com/TheRockStarDBA/c48912abfcc176cf77b0 ?

**Comment by marcingminski** [comment]
I like that. I will look into it. Thank you.
